sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala

```diff
@@ -175,6 +178,98 @@ case class AlterTableRenameCommand(
 }
 
 /**
+ * A command that add columns to a table
+ * The syntax of using this command in SQL is:
+ * {{{
+ *   ALTER TABLE table_identifier
+ *   ADD COLUMNS (col_name data_type [COMMENT col_comment], ...);
+ * }}}
+*/
+case class AlterTableAddColumnsCommand(
+    table: TableIdentifier,
+    columns: Seq[StructField]) extends RunnableCommand {
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val catalog = sparkSession.sessionState.catalog
+    val catalogTable = verifyAlterTableAddColumn(catalog, table)
+
+    try {
+      sparkSession.catalog.uncacheTable(table.quotedString)
+    } catch {
+      case NonFatal(e) =>
+        log.warn(s"Exception when attempting to uncache table ${table.quotedString}", e)
+    }
+
+    // Invalidate the table last, otherwise uncaching the table would load the logical plan
+    // back into the hive metastore cache
+    catalog.refreshTable(table)
+    val partitionFields = catalogTable.schema.takeRight(catalogTable.partitionColumnNames.length)
+    val newSchemaFields = catalogTable.schema
+      .take(catalogTable.schema.length - catalogTable.partitionColumnNames.length) ++
+      columns ++ partitionFields
+    checkDuplication(sparkSession, newSchemaFields)
```

Comment:
let's move the partition column arrangement and the duplication check into `SessionCatalog.alterTableSchema`